pkg/proxy/util/nodeport_addresses.go

```diff
@@ -35,27 +36,45 @@ type NodePortAddresses struct {
 // RFC 5735 127.0.0.0/8 - This block is assigned for use as the Internet host loopback address
 var ipv4LoopbackStart = net.IPv4(127, 0, 0, 0)
 
-// NewNodePortAddresses takes the `--nodeport-addresses` value (which is assumed to
-// contain only valid CIDRs) and returns a NodePortAddresses object. If cidrStrings is
-// empty, this is treated as `["0.0.0.0/0", "::/0"]`.
-func NewNodePortAddresses(cidrStrings []string) *NodePortAddresses {
-	if len(cidrStrings) == 0 {
-		cidrStrings = []string{IPv4ZeroCIDR, IPv6ZeroCIDR}
+// NewNodePortAddresses takes an IP family and the `--nodeport-addresses` value (which is
+// assumed to contain only valid CIDRs, potentially of both IP families) and returns a
+// NodePortAddresses object for the given family. If there are no CIDRs of the given
+// family then the CIDR "0.0.0.0/0" or "::/0" will be added (even if there are CIDRs of
+// the other family).
+func NewNodePortAddresses(family v1.IPFamily, cidrStrings []string) *NodePortAddresses {
+	npa := &NodePortAddresses{}
+
+	// Filter CIDRs to correct family
+	for _, str := range cidrStrings {
+		if (family == v1.IPv4Protocol) == netutils.IsIPv4CIDRString(str) {
+			npa.cidrStrings = append(npa.cidrStrings, str)
+		}
 	}
-
-	npa := &NodePortAddresses{
-		cidrStrings: cidrStrings,
+	if len(npa.cidrStrings) == 0 {
+		if family == v1.IPv4Protocol {
+			npa.cidrStrings = []string{IPv4ZeroCIDR}
+		} else {
+			npa.cidrStrings = []string{IPv6ZeroCIDR}
+		}
 	}
 
+	// Now parse
 	for _, str := range npa.cidrStrings {
 		_, cidr, _ := netutils.ParseCIDRSloppy(str)
-		npa.cidrs = append(npa.cidrs, cidr)
 
 		if netutils.IsIPv4CIDR(cidr) {
 			if cidr.IP.IsLoopback() || cidr.Contains(ipv4LoopbackStart) {
 				npa.containsIPv4Loopback = true
 			}
 		}
+
+		if IsZeroCIDR(str) {
```

Comment:
why don't you return directly in line 53 with the zero cidr instead of parsing it twice ?

Reply:
It's not clear to me exactly which line 53 you are talking about, but if you're asking about the `if len(npa.cidrStrings) == 0` case, then remember that have to deal with both `cidrStrings = []` and `cidrStrings = ["0.0.0.0"]`, so we can't _move_ any code from here to that case, we'd have to duplicate it in both cases.

Also, `IsZeroCIDR` doesn't actually parse `str`, it just compares it against two constants...

Reply:
yeah, make sense, I was going commit by commit but once you see all together is more evident